### DISCUSSION TITLE
Is there a way to save TGraph to a root file using uproot ?

### DISCUSSION THREAD

**Discussion by grzanka** [scikit-hep/uproot5#1142] [Q&A]
I would like to perform similar operation as here, but using uproot:
https://github.com/scikit-hep/uproot5/blob/1f0e557976b902b8de749b17b3aaa516c7ae52c1/dev/example-objects.py#L157

Namely to create TGraph and save it in the ROOT file:
```
import numpy as np
import ROOT

file = ROOT.TFile("example-objects.root", "recreate")

x = np.array([1.1, 2.2, 3.3, 4.4, 5.5])
y = np.array([1.0, -2.0, 3.0, -4.0, 5.0])

tgraph = ROOT.TGraph(5, x, y)
tgraph.SetName("tgraph")
tgraph.SetTitle("title")
tgraph.Write()

file.Close()
```

I saw some discussion here: https://github.com/scikit-hep/uproot5/issues/309 but it seemed to be related to reading

**Comment by jpivarski** [comment]
This comment started off as an explanation of why this would be hard to do, but along the way, I found that the job was half-done for some reason. There wasn't much left to be done, so writing TGraph is a plausible feature request.

---------

Reading and writing are very different. Even before #309, Uproot could read TGraphs because the TStreamerInfo in the ROOT file specified how the sequence of bytes is to be interpreted as a data structure with attributes; #309 added methods that interpret those attributes in a Pythonic way.

To write a data type, the data-attributes-to-bytes layout has to be known to Uproot by explicit implementation, not discovered from a file. We'd also need to add TStreamerInfo for the TGraph class, its superclasses, and any attributes that are class types. We've done this for TObjString, all of the histogram types, TTree and TBranch, and all of the TLeaf types, so that we can write strings, histograms, and TTrees.

I was going to point out where these things would need to be added for TGraph, and interestingly, some of them are already there. We've included a copy of TGraph's TStreamerInfo:

https://github.com/scikit-hep/uproot5/blob/1f0e557976b902b8de749b17b3aaa516c7ae52c1/src/uproot/models/TGraph.py#L22-L27

(we do this as raw bytes, without interpreting the content of those bytes, for simplicity; that's why it's a raw dump, copied from a ROOT file). However, the `serialize` methods for TGraph, TGraphErrors, and TGraphAsymmErrors are all not filled in; they're blocked with a `raise NotImplementedError("FIXME")`:

https://github.com/scikit-hep/uproot5/blob/1f0e557976b902b8de749b17b3aaa516c7ae52c1/src/uproot/models/TGraph.py#L337-L345

Using your example-objects.root to see what classes would need to be writable to write a TGraph, it looks like the only new thing that we haven't implemented yet is TGraph itself (not any hidden superclasses or nested classes):

```python
>>> f = uproot.open("example-objects.root")
>>> f.file.show_streamers("TGraph")
TArrayF (v1): TArray (v1)
    fArray: float* (TStreamerBasicPointer)

THashList (v0): TList (v5)

TArray (v1)
    fN: int (TStreamerBasicType)

TArrayD (v1): TArray (v1)
    fArray: double* (TStreamerBasicPointer)

TAttAxis (v4)
    fNdivisions: int (TStreamerBasicType)
    fAxisColor: short (TStreamerBasicType)
    fLabelColor: short (TStreamerBasicType)
    fLabelFont: short (TStreamerBasicType)
    fLabelOffset: float (TStreamerBasicType)
    fLabelSize: float (TStreamerBasicType)
    fTickLength: float (TStreamerBasicType)
    fTitleOffset: float (TStreamerBasicType)
    fTitleSize: float (TStreamerBasicType)
    fTitleColor: short (TStreamerBasicType)
    fTitleFont: short (TStreamerBasicType)

TAxis (v10): TNamed (v1), TAttAxis (v4)
    fNbins: int (TStreamerBasicType)
    fXmin: double (TStreamerBasicType)
    fXmax: double (TStreamerBasicType)
    fXbins: TArrayD (TStreamerObjectAny)
    fFirst: int (TStreamerBasicType)
    fLast: int (TStreamerBasicType)
    fBits2: unsigned short (TStreamerBasicType)
    fTimeDisplay: bool (TStreamerBasicType)
    fTimeFormat: TString (TStreamerString)
    fLabels: THashList* (TStreamerObjectPointer)
    fModLabs: TList* (TStreamerObjectPointer)

TH1 (v8): TNamed (v1), TAttLine (v2), TAttFill (v2), TAttMarker (v2)
    fNcells: int (TStreamerBasicType)
    fXaxis: TAxis (TStreamerObject)
    fYaxis: TAxis (TStreamerObject)
    fZaxis: TAxis (TStreamerObject)
    fBarOffset: short (TStreamerBasicType)
    fBarWidth: short (TStreamerBasicType)
    fEntries: double (TStreamerBasicType)
    fTsumw: double (TStreamerBasicType)
    fTsumw2: double (TStreamerBasicType)
    fTsumwx: double (TStreamerBasicType)
    fTsumwx2: double (TStreamerBasicType)
    fMaximum: double (TStreamerBasicType)
    fMinimum: double (TStreamerBasicType)
    fNormFactor: double (TStreamerBasicType)
    fContour: TArrayD (TStreamerObjectAny)
    fSumw2: TArrayD (TStreamerObjectAny)
    fOption: TString (TStreamerString)
    fFunctions: TList* (TStreamerObjectPointer)
    fBufferSize: int (TStreamerBasicType)
    fBuffer: double* (TStreamerBasicPointer)
    fBinStatErrOpt: TH1::EBinErrorOpt (TStreamerBasicType)
    fStatOverflows: TH1::EStatOverflows (TStreamerBasicType)

TH1F (v3): TH1 (v8), TArrayF (v1)

TCollection (v3): TObject (v1)
    fName: TString (TStreamerString)
    fSize: int (TStreamerBasicType)

TSeqCollection (v0): TCollection (v3)

TList (v5): TSeqCollection (v0)

TAttMarker (v2)
    fMarkerColor: short (TStreamerBasicType)
    fMarkerStyle: short (TStreamerBasicType)
    fMarkerSize: float (TStreamerBasicType)

TAttFill (v2)
    fFillColor: short (TStreamerBasicType)
    fFillStyle: short (TStreamerBasicType)

TAttLine (v2)
    fLineColor: short (TStreamerBasicType)
    fLineStyle: short (TStreamerBasicType)
    fLineWidth: short (TStreamerBasicType)

TString (v2)

TObject (v1)
    fUniqueID: unsigned int (TStreamerBasicType)
    fBits: unsigned int (TStreamerBasicType)

TNamed (v1): TObject (v1)
    fName: TString (TStreamerString)
    fTitle: TString (TStreamerString)

TGraph (v4): TNamed (v1), TAttLine (v2), TAttFill (v2), TAttMarker (v2)
    fNpoints: int (TStreamerBasicType)
    fX: double* (TStreamerBasicPointer)
    fY: double* (TStreamerBasicPointer)
    fFunctions: TList* (TStreamerObjectPointer)
    fHistogram: TH1F* (TStreamerObjectPointer)
    fMinimum: double (TStreamerBasicType)
    fMaximum: double (TStreamerBasicType)
```

All of the classes that TGraph depends on are histogram-related, which are already implemented in Uproot, and anyway they wouldn't matter if we only write TGraph without an attached `fHistogram` (i.e. set it to `nullptr`). The serialization of the TGraph itself consists of only

* `fNpoints`, a 32-bit integer
* `fX`, a basic array of `64-bit floats
* `fY`, a basic array of 64-bit floats
* `fFunctions`, which we can set to `nullptr`
* `fHistogram`, which we can set to `nullptr`
* `fMinimum`, a 64-bit float
* `fMaximum`, a 64-bit float

It could, perhaps, be a single [struct.pack](https://docs.python.org/3/library/struct.html#struct.pack) command.

I put a `cursor.debug(chunk)` statement at the top of [Model_TGraph_v4.read_members](https://github.com/scikit-hep/uproot5/blob/1f0e557976b902b8de749b17b3aaa516c7ae52c1/src/uproot/models/TGraph.py#L37) and read the TGraph from example-objects.root. These are the raw bytes (printed as decimal, not hexidecimal) and character interpretations wherever possible:

```
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 64   0   0  25   0   1   0   1   0   0   0   0   0   0   4   0   6 116 103 114
  @ --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---   t   g   r
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 97 112 104   5 116 105 116 108 101  64   0   0   8   0   2   0   1   0   1   0
  a   p   h ---   t   i   t   l   e   @ --- --- --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  1  64   0   0   6   0   2   0   0   3 232  64   0   0  10   0   2   0   1   0
---   @ --- --- --- --- --- --- --- --- ---   @ --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  1  63 128   0   0   0   0   0   5   1  63 241 153 153 153 153 153 154  64   1
---   ? --- --- --- --- --- --- --- ---   ? --- --- --- --- --- --- ---   @ ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
153 153 153 153 153 154  64  10 102 102 102 102 102 102  64  17 153 153 153 153
--- --- --- --- --- ---   @ ---   f   f   f   f   f   f   @ --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
153 154  64  22   0   0   0   0   0   0   1  63 240   0   0   0   0   0   0 192
--- ---   @ --- --- --- --- --- --- --- ---   ? --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  0   0   0   0   0   0   0  64   8   0   0   0   0   0   0 192  16   0   0   0
--- --- --- --- --- --- ---   @ --- --- --- --- --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  0   0   0  64  20   0   0   0   0   0   0  64   0   0  31 255 255 255 255  84
--- --- ---   @ --- --- --- --- --- --- ---   @ --- --- --- --- --- --- ---   T
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 76 105 115 116   0  64   0   0  17   0   5   0   1   0   0   0   0   0   0   0
  L   i   s   t ---   @ --- --- --- --- --- --- --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  0   0   0   0   0   0   0   0   0   0 192 145  92   0   0   0   0   0 192 145
--- --- --- --- --- --- --- --- --- --- --- ---   \ --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 92   0   0   0   0   0
  \ --- --- --- --- ---
```

The fields are

* `fNpoints`: 5
* `fX`: array of length 5, `1.1, 2.2, 3.3, 4.4, 5.5`
* `fY`: array of length 5, `1., -2., 3., -4., 5.`
* `fFunctions`: TList of zero items (okay, not a `nullptr`)
* `fHistogram`: a `nullptr`
* `fMinimum`: -1111.0
* `fMaximum`: -1111.0

Using

```python
>>> [int(x) for x in np.array([1.1, 2.2, 3.3, 4.4, 5.5], dtype='>f8').tobytes()]
[63, 241, 153, 153, 153, 153, 153, 154, 64, 1, 153, 153, 153, 153, 153, 154, 64, 10, 102, 102, 102, 102, 102, 102, 64, 17, 153, 153, 153, 153, 153, 154, 64, 22, 0, 0, 0, 0, 0, 0]
```

to find where `fX` is by eye, it looks like this part:

```
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 64   0   0  25   0   1   0   1   0   0   0   0   0   0   4   0   6 116 103 114
  @ --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---   t   g   r
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
 97 112 104   5 116 105 116 108 101  64   0   0   8   0   2   0   1   0   1   0
  a   p   h ---   t   i   t   l   e   @ --- --- --- --- --- --- --- --- --- ---
--+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+-
  1  64   0   0   6   0   2   0   0   3 232  64   0   0  10   0   2   0   1   0
---   @ --- --- --- --- --- --- --- --- ---   @ --- --- --- --- --- --- --- ---
--+---+---+---+---+
  1  63 128   0   0
---   ? --- --- ---
```

is all of the superclasses, which the `serialize` implementation already includes, and the part that's currently `raise NotImplementedError("FIXME")` is a `struct.pack` like this:

```python
>>> expected_bytes = np.array([0, 0, 0, 5, 1, 63, 241, 153, 153, 153, 153, 153, 154, 64, 1, 153, 153, 153, 153, 153, 154, 64, 10, 102, 102, 102, 102, 102, 102, 64, 17, 153, 153, 153, 153, 153, 154, 64, 22, 0, 0, 0, 0, 0, 0, 1, 63, 240, 0, 0, 0, 0, 0, 0, 192, 0, 0, 0, 0, 0, 0, 0, 64, 8, 0, 0, 0, 0, 0, 0, 192, 16, 0, 0, 0, 0, 0, 0, 64, 20, 0, 0, 0, 0, 0, 0, 64, 0, 0, 31, 255, 255, 255, 255, 84, 76, 105, 115, 116, 0, 64, 0, 0, 17, 0, 5, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 192, 145, 92, 0, 0, 0, 0, 0, 192, 145, 92, 0, 0, 0, 0, 0], np.uint8)
>>> expected_bytes.tobytes()
b'\x00\x00\x00\x05\x01?\xf1\x99\x99\x99\x99\x99\x9a@\x01\x99\x99\x99\x99\x99\x9a@\nffffff@\x11\x99\x99\x99\x99\x99\x9a@\x16\x00\x00\x00\x00\x00\x00\x01?\xf0\x00\x00\x00\x00\x00\x00\xc0\x00\x00\x00\x00\x00\x00\x00@\x08\x00\x00\x00\x00\x00\x00\xc0\x10\x00\x00\x00\x00\x00\x00@\x14\x00\x00\x00\x00\x00\x00@\x00\x00\x1f\xff\xff\xff\xffTList\x00@\x00\x00\x11\x00\x05\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc0\x91\\\x00\x00\x00\x00\x00\xc0\x91\\\x00\x00\x00\x00\x00'
>>>
>>> attempt = struct.pack(">i", 5) + b"\x01" + np.array([1.1, 2.2, 3.3, 4.4, 5.5], dtype=">f8").tobytes() + b"\x01" + np.array([1., -2., 3., -4., 5.], dtype=">f8").tobytes() + b"@\x00\x00\x1f\xff\xff\xff\xffTList\x00@\x00\x00\x11\x00\x05\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00" + struct.pack(">d", -1111.0) + struct.pack(">d", -1111.0)
>>> attempt
b'\x00\x00\x00\x05\x01?\xf1\x99\x99\x99\x99\x99\x9a@\x01\x99\x99\x99\x99\x99\x9a@\nffffff@\x11\x99\x99\x99\x99\x99\x9a@\x16\x00\x00\x00\x00\x00\x00\x01?\xf0\x00\x00\x00\x00\x00\x00\xc0\x00\x00\x00\x00\x00\x00\x00@\x08\x00\x00\x00\x00\x00\x00\xc0\x10\x00\x00\x00\x00\x00\x00@\x14\x00\x00\x00\x00\x00\x00@\x00\x00\x1f\xff\xff\xff\xffTList\x00@\x00\x00\x11\x00\x05\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc0\x91\\\x00\x00\x00\x00\x00\xc0\x91\\\x00\x00\x00\x00\x00'
>>> attempt == expected_bytes.tobytes()
True
```

So I tried an implementation:

```diff
--- a/src/uproot/models/TGraph.py
+++ b/src/uproot/models/TGraph.py
@@ -338,7 +338,16 @@ class Model_TGraph_v4(uproot.behaviors.TGraph.TGraph, uproot.model.VersionedMode
         where = len(out)
         for x in self._bases:
             x._serialize(out, True, name, tobject_flags)
-        raise NotImplementedError("FIXME")
+        out.extend([
+            struct.pack(">i", self._members["fNpoints"]),
+            b"\x01",
+            self._members["fX"].astype(">f8").tobytes(),
+            b"\x01",
+            self._members["fY"].astype(">f8").tobytes(),
+            b"@\x00\x00\x1f\xff\xff\xff\xffTList\x00@\x00\x00\x11\x00\x05\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00",
+            struct.pack(">d", self._members["fMinimum"]),
+            struct.pack(">d", self._members["fMaximum"]),
+        ])
         if header:
             num_bytes = sum(len(x) for x in out[where:])
             version = 4
```

and it can convert the TGraph into a PyROOT object, which is a demonstration that serialization works (since that's how we convert objects between Uproot and PyROOT):

```python
>>> f = uproot.open("/tmp/example-objects.root")
>>> g = f["tgraph"]
>>> g2 = uproot.to_pyroot(g)
>>> type(g2)
<class cppyy.gbl.TGraph at 0x600946035810>
>>> g2.GetName()
'tgraph'
>>> g2.GetTitle()
'title'
>>> np.asarray(g2.GetX())
array([1.1, 2.2, 3.3, 4.4, 5.5])
>>> np.asarray(g2.GetY())
array([ 1., -2.,  3., -4.,  5.])
>>> g2.GetMinimum()
-1111.0
>>> g2.GetMaximum()
-1111.0
```

The above does _not_ test whether the streamers are being handled correctly, but it appears as though they have a correct implementation already. That would just need to be tested.

[src/uproot/writing/identify.py](https://github.com/scikit-hep/uproot5/blob/main/src/uproot/writing/identify.py) would need an entry point for syntax like

```python
output_file["name"] = XYZ
```

to work (and as discussed on #1144, we'd have to decide what an appropriate Python object XYZ is, to be interpreted as a TGraph).

---------------

This started out as a comment saying why this project would be hard, but since half of it was already implemented, it's not taking much to get the implementation finished. Could you take over from here? Do you have enough context from what I've started? I can help, and I can give you permissions to edit the `jpivarski/start-implementation-of-TGraph-writing` branch in PR #1144.

**Reply by grzanka** [reply]
Hi, 

You gave a very comprehensive overview, that is really a lot for a starting point. I think I understood (and got the tools to debug) the streamer part.

As for "write by assignment" of TGraph I have no good ideas at the moment. The only thing which comes to my mind is to use assigment by a method, something similar to `file.mktree`. You could then call for example:

```python
output_file.mkgraph("my_graph", x=[1.1, 2.2, 3.3, 4.4, 5.5], y=[ 1., -2.,  3., -4.,  5.])
```

The exact signature could be deduced from ROOT TGraph constructor and their default values.

As for `identify.py` the logic of `to_writable` seems pretty complex, but if there is a good representation of TGraph in python object, I think it needs to go there as well:
https://github.com/scikit-hep/uproot5/blob/d2464242b1545a81b30bfd7413e25102bb6650c8/src/uproot/writing/identify.py#L185

As for my questions - what is the role of functions like this:
https://github.com/scikit-hep/uproot5/blob/d2464242b1545a81b30bfd7413e25102bb6650c8/src/uproot/writing/identify.py#L1600C5-L1600C16
? Is it only for internal usage ? I think TGraph maybe added there as well so the testing would be easier.

I take a look on the PR at some point next week.

**Comment by grzanka** [comment]
Also - work on this would help me to understand the issues in https://github.com/scikit-hep/uproot5/discussions/1135#discussioncomment-8557000

**Comment by grzanka** [comment]
@jpivarski where do you prefer to handle the discussion on possible python object to interpret as TGraph ? here or in the PR https://github.com/scikit-hep/uproot5/pull/1144#issue-2151558399 ?

I see another point that some of the propositions may cause confusions for users. Let's assume we go with DataFrame with `x` and `y` columns. User may would like to save it as TTree, as documentation suggest here:
<img width="725" alt="Screenshot 2024-02-26 at 10 44 30" src="https://github.com/scikit-hep/uproot5/assets/7374733/8ae1d748-5c11-4118-b013-d036696548fd">
or as TGraph as we discuss above. 

The basic data behind TGraph (x and y) seems like basic ingredient for TTree as well. The "write by assignment" doesn't give many options to specify the type in the ROOT file to be saved.

Maybe if we require some other ingredient which makes easier to deduce a datatype, for example DataFrame attribute which is unique to TGraph and doesn't make sense to TTree ?

**Reply by jpivarski** [reply]
It would probably be better to discuss this on the PR.

You're right! A DataFrame is effectively like a dict of NumPy arrays, and therefore it's (rightly) interpreted as a TTree. We can't have TGraph and TTree overlap.

One thing that's very different about a TGraph is that so few column names are possible. A TGraph can have `x` and `y`, TGraphErrors can have those and also `xerr` and `yerr`, and a TGraphAsymmErrors can have two more arguments for asymmetric errors. A DataFrame (and a dict) can have arbitrarily many columns with arbitrary names. Maybe instead of something with `__getitem__` access (`df["x"]`), it should be something with `__getattr__` access (`obj.x` and `obj.y`). If [uproot.writing.to_writable](https://github.com/scikit-hep/uproot5/blob/ea207f992c61bd2265e112f29c8b0c3d0cd6706d/src/uproot/writing/identify.py#L185) is given an object with these specific attributes, it should interpret them as a TGraph?

There's still a danger of overlap here: Pandas DataFrames allow both `__getitem__` and `__getattr__` access, so if your DataFrame has fields named `x` and `y`, it would be interpreted as a TGraph, and if it has any other names, it would be interpreted as a TTree. Also, namedtuples are both tuples and have attributes, so they could be interpreted as both a histogram and a TGraph. It would be better—less potential for user surprise—if there was no overlap, if that which is interpreted as a TGraph is a completely different _type_ from anything that can be interpreted as a TTree or a histogram.
